src/Debugging/Microsoft.DiaSymReader.PortablePdb/SymBinder.cs

```diff
@@ -11,106 +17,357 @@ namespace Microsoft.DiaSymReader.PortablePdb
 
-            byte[] bytes;
-            try
+            Guid guid = codeViewData.PdbId;
+            int age = codeViewData.Age;
+            var metadataImportProvider = new Func<IMetadataImport>(() => mdImport);
+            string pdbFileName = Path.GetFileName(codeViewData.Path);
+
+            // 1) next to the PE file 
+            if ((searchPolicy & SymUnmanagedSearchPolicy.AllowReferencePathAccess) != 0)
             {
-                // TODO: use memory mapped files?
-                bytes = PortableShim.File.ReadAllBytes(fileName);
+                string peDirectory = Path.GetDirectoryName(fileName);
+                string pdbFilePath = Path.Combine(peDirectory, pdbFileName);
+
+                if (TryCreateReaderForMatchingPdb(pdbFilePath, guid, stamp, age, metadataImportProvider, out reader))
+                {
+                    return HResult.S_OK;
+                }
             }
-            catch
+
+            // 2) PDB path as specified in Debug Directory
+            if ((searchPolicy & SymUnmanagedSearchPolicy.AllowOriginalPathAccess) != 0)
             {
-                reader = null;
-                return HResult.E_INVALIDARG;
+                if (TryCreateReaderForMatchingPdb(codeViewData.Path, guid, stamp, age, metadataImportProvider, out reader))
+                {
+                    return HResult.S_OK;
+                }
+            }
+
+            // 3) Search Paths
+            string peFileExtension = Path.GetExtension(fileName).TrimStart('.');
+
+            foreach (var searchPaths in GetSearchPathsSequence(searchPath, searchPolicy))
+            {
+                if (TryFindMatchingPdb(searchPaths, peFileExtension, pdbFileName, guid, stamp, age, metadataImportProvider, searchPolicy, out reader))
+                {
+                    return HResult.S_OK;
+                }
             }
 
-            reader = new SymReader(new PortablePdbReader(bytes, bytes.Length, metadataImport));
-            return HResult.S_OK;
+            return HResult.E_PDB_NOT_FOUND; 
         }
 
-        [PreserveSig]
-        public int GetReaderFromStream(
-            [MarshalAs(UnmanagedType.Interface)]object importer,
-            [MarshalAs(UnmanagedType.Interface)]object stream,
-            [MarshalAs(UnmanagedType.Interface)]out ISymUnmanagedReader reader)
+        private static IEnumerable<string> GetSearchPathsSequence(string searchPath, SymUnmanagedSearchPolicy searchPolicy)
         {
-            if (importer == null || stream == null)
+            // 3a) parameter
+            yield return searchPath;
+
+            // 4b) registry
+            if ((searchPolicy & SymUnmanagedSearchPolicy.AllowRegistryAccess) != 0)
             {
-                reader = null;
-                return HResult.E_INVALIDARG;
+                // TODO
+            }
+
+            // 5c) environment variables:
+            yield return PortableShim.Environment.GetEnvironmentVariable("_NT_ALT_SYMBOL_PATH");
```

Comment:
Should you have an 'if windows' check for these? Do you only want to yield return them if they are non-null?